File: server/src/main/java/bgu/spl/net/impl/stomp/Backend/Frames/DisconnectFrame.java
package bgu.spl.net.impl.stomp.Backend.Frames;

import java.io.IOException;
import java.util.HashMap;

import bgu.spl.net.genericServers.interfaces.ConnectionHandler;

public class DisconnectFrame extends ExecutableFrame {

    private static String RECEIPT_FIELD = "receipt";

    protected DisconnectFrame(HashMap<String,String> headers, String frameBody) {
        super(headers, frameBody, StompCommand.DISCONNECT);
    }

    @Override
    public String execute(ConnectionHandler<String> handler) {
        
        int receipt = Integer.parseInt(headers.get(RECEIPT_FIELD));

        try{
            conM.disconnect(handler);
            handler.close();
            return ReceiptFrame.get(receipt).toString();
        }catch (IOException e){
            return ErrorFrame.get(e.getMessage()).toString();
        }
    }
  
}

File: client/src/frames/DisconnectFrame.cpp
#include "DisconnectFrame.h"
#include "UserData.h"

DisconnectFrame::DisconnectFrame(unordered_map<string, string> headers, string frameBody)
    : Frame(StompCommand::DISCONNECT, headers, frameBody){}


DisconnectFrame * DisconnectFrame::get()
{
    unordered_map<string, string> headersMap;
    headersMap["receipt"] = to_string(UserData::getInstance().getReceiptId());
    DisconnectFrame *frame = new DisconnectFrame(headersMap, "");
    return frame;
}

